darwin/DarwinProcessList.c

```diff
@@ -128,13 +133,81 @@ static struct kinfo_proc* ProcessList_getKInfoProcs(size_t* count) {
    CRT_fatalError("Unable to get kinfo_procs");
 }
 
+#ifdef HAVE_POWER_GADGET
+// in order to enable separate distribution, we link weakly with the
+// framework
+extern bool PG_Initialize(void) __attribute__((weak_import));
+
+static void DarwinProcessList_scanCPUFrequency(DarwinProcessList* this) {
+   int cpus = this->super.cpuCount;
+
+   for (int i = 0; i < cpus; i++) {
+      this->cpu_freqs[i] = NAN;
+      this->cpu_temps[i] = NAN;
+   }
+
+   if (!this->power_gadget_is_available && PG_Initialize != NULL) {
+      this->power_gadget_is_available = PG_Initialize();
+   }
+
+   if (this->power_gadget_is_available) {
+      int package_count = 0;
+      int core_idx = 0;
+
+      if (PG_GetNumPackages(&package_count)) {
+         for (int p = 0; p < package_count; p++) {
+            PGSampleID new_sample;
+
+            if (PG_ReadSample(p, &new_sample)) {
+               if (this->cpu_sample) {
+                  int core_count = 0;
+
+                  if (PG_GetNumCores(p, &core_count)) {
+                     for (int c = 0; c < core_count && core_idx < cpus; c++) {
+                        double mean, min, max;
+
+                        if (PGSample_GetIACoreFrequency(this->cpu_sample,
+                                                        new_sample, c,
+                                                        &mean, &min, &max)) {
+                           this->cpu_freqs[core_idx] = mean;
+                        }
+
+                        if (PGSample_GetIACoreTemperature(this->cpu_sample, c,
+                                                          &mean, &min, &max)) {
+                           this->cpu_temps[core_idx] = mean;
+                        }
+
+                        core_idx += 1;
+                     }
+                  }
+
+                  PGSample_Release(this->cpu_sample);
+               }
+
+               this->cpu_sample = new_sample;
+            }
+         }
+      }
+   }
```

Comment:
Try to reduce the nesting by e.g. inverting conditions and possibly early exiting.